### ISSUE TITLE
Static Seed Auto Discovery Seems Unstable in Web Dashboard

### ISSUE THREAD

**Issue by seinshah** [emqx/emqx#12321]
### What happened?

I am deploying EMQX on AWS EC2 and use static method of auto clustering. All Ec2 instances are allowed to communicate with each other on all ports.

- Created 1 core and 2 replicant nodes in the beginning
- Dashboard showed these without any issue
- Added 1 more core and 2 more replicant node
- Dashboard struggles to show these new nodes as running constantly. It keeps switching these nodes from running to stopped and vice-versa
- Running `emqx_ctl cluster status` repeatedly show all nodes as running without problem
- When a client connects to one of these on/off nodes, dashboard also struggles to show the connection number properly and keeps changing from X to X - 1 and vice-versa

For instance in the following image, it shows the second core node as stopped, but it immediately turns green. The interval is even less than a second:

![Screenshot 2024-01-12 at 13 27 11](https://github.com/emqx/emqx/assets/3847935/56737613-9769-4fba-b2a9-eaf128da03a4)

### What did you expect to happen?

Expected to see all the newly added core and replicant nodes to be shown as always running on the dashboard and the number of live connections reflect the correct number of e.g. 2 instead of changing from 1 to 2 and vice-versa

### How can we reproduce it (as minimally and precisely as possible)?

<details>
Running on AWS EC2 with the following AMI

```
data "aws_ami" "ubuntu" {
  most_recent = true
  owners      = ["099720109477"]

  filter {
    name   = "name"
    values = ["ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-*"]
  }

  filter {
    name   = "root-device-type"
    values = ["ebs"]
  }

  filter {
    name   = "virtualization-type"
    values = ["hvm"]
  }
}
```
</details>

### EMQX version

<details>

```console
sysdescr  : EMQX
version   : 5.4.1
datetime  : 2024-01-12T12:30:30.901203581+00:00
uptime    : 1 hours, 1 minutes, 22 seconds
```

</details>

### OS version

<details>

```console
NAME="Ubuntu"
VERSION="20.04.6 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.6 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

</details>

### Log files

<details>

</details>

**Comment by ieQu1** [comment]
Hello,

I suspect this is might be a dashboard problem. It would be helpful for us to see:

1. `emqx.log.x` files from the EMQX nodes
2. Is there any warnings in the browser debug console?

**Comment by seinshah** [comment]
Thanks for the response @ieQu1.
The following was the only message in `emqx.log.*`

```
2024-01-12T11:41:47.804168+00:00 [warning] msg: dashboard_monitor_error, mfa: emqx_dashboard_monitor:current_rate/1(144), reason: {noproc,{gen_server,call,[emqx_dashboard_monitor,current_rate,5000]}}
```

I have removed the resources for now, but if `warnings in the browser debug console` is necessary, please let me know to re-deploy and monitor the situation.

**Comment by ieQu1** [comment]
It can't be the only message in the logs... Normally EMQX outputs quite a few messages when it joins the cluster. Can I see full logs (not just warnings)?

**Comment by seinshah** [comment]
Here are the logs (`cat emqx/log/*`) from each of the nodes on the custer:

**TL;DR**

- There is a warning with `reason: join` in the second core node initialization
- There is no log in any of the nodes regarding joining the cluster
- First replicant shows the second core node as `stopped` while other nodes claim it's `running`
- Second core does not list the second replicant as part of the cluster nodes, neither `running` nor `stopped`

---

**First Code Node** (`emqx@10.1.1.147`)
This was the first core node and therefore no other node existed to join to.

Result of `emqx_ctl cluster status`:

```
Cluster status: #{running_nodes =>
                      ['emqx@10.1.1.114','emqx@10.1.1.147','emqx@10.1.2.152',
                       'emqx@10.1.2.222'],
                  stopped_nodes => []}
```

Logs:  

<details>

```
=====
===== LOGGING STARTED Wed Jan 17 10:59:11 UTC 2024
=====
Erlang/OTP 25 [erts-13.2.2] [emqx] [64-bit] [smp:2:2] [ds:2:2:8] [async-threads:4] [jit:ns]

Listener ssl:default on 0.0.0.0:8883 started.
Listener tcp:default on 0.0.0.0:1883 started.
Listener ws:default on 0.0.0.0:8083 started.
Listener wss:default on 0.0.0.0:8084 started.
Listener http:dashboard on :18083 started.
EMQX 5.4.1 is running now!
Restricted Eshell V13.2.2  (abort with ^G)
v5.4.1(emqx@10.1.1.147)1> run_erl [2358] Wed Jan 17 10:59:11 2024
Args before exec of shell:
run_erl [2358] Wed Jan 17 10:59:11 2024
argv[0] = sh
run_erl [2358] Wed Jan 17 10:59:11 2024
argv[1] = -c
run_erl [2358] Wed Jan 17 10:59:11 2024
argv[2] = exec "/home/ubuntu/emqx/bin/emqx" "console"
```

</details>

---

**First Replicant Node** (`emqx@10.1.1.114`)
This was created immediately after creating the first core node and joined the first core node as provided in the static seed.

Result of `emqx_ctl cluster status`:

```
Cluster status: #{running_nodes =>
                      ['emqx@10.1.1.114','emqx@10.1.1.147','emqx@10.1.2.152'],
                  stopped_nodes => ['emqx@10.1.2.222']}
```

Logs:  

<details>

```
=====
===== LOGGING STARTED Wed Jan 17 10:59:50 UTC 2024
=====
Erlang/OTP 25 [erts-13.2.2] [emqx] [64-bit] [smp:2:2] [ds:2:2:8] [async-threads:4] [jit:ns]

Listener ssl:default on 0.0.0.0:8883 started.
Listener tcp:default on 0.0.0.0:1883 started.
Listener ws:default on 0.0.0.0:8083 started.
Listener wss:default on 0.0.0.0:8084 started.
Listener http:dashboard on :18083 started.
EMQX 5.4.1 is running now!
Restricted Eshell V13.2.2  (abort with ^G)
v5.4.1(emqx@10.1.1.114)1>
===== ALIVE Wed Jan 17 11:15:03 UTC 2024
run_erl [2351] Wed Jan 17 10:59:50 2024
Args before exec of shell:
run_erl [2351] Wed Jan 17 10:59:50 2024
argv[0] = sh
run_erl [2351] Wed Jan 17 10:59:50 2024
argv[1] = -c
run_erl [2351] Wed Jan 17 10:59:50 2024
argv[2] = exec "/home/ubuntu/emqx/bin/emqx" "console"
```

</details>

---

**Second Core Node** (`emqx@10.1.2.222`)
This was added later to the cluster with first core and replicant nodes as the static seeds.

Result of `emqx_ctl cluster status`:

```
Cluster status: #{running_nodes =>
                      ['emqx@10.1.1.147','emqx@10.1.2.152','emqx@10.1.2.222'],
                  stopped_nodes => []}
```

Logs:  

<details>

```
=====
===== LOGGING STARTED Wed Jan 17 11:07:42 UTC 2024
=====
Erlang/OTP 25 [erts-13.2.2] [emqx] [64-bit] [smp:2:2] [ds:2:2:8] [async-threads:4] [jit:ns]

Listener ssl:default on 0.0.0.0:8883 started.
Listener tcp:default on 0.0.0.0:1883 started.
Listener ws:default on 0.0.0.0:8083 started.
Listener wss:default on 0.0.0.0:8084 started.
Listener http:dashboard on :18083 started.
EMQX 5.4.1 is running now!
Restricted Eshell V13.2.2  (abort with ^G)
v5.4.1(emqx@10.1.2.222)1> 2024-01-17T11:07:54.605274+00:00 [warning] msg: Stopping mria, mfa: mria:stop/1(134), reason: join
Stop listener http:dashboard on :18083 successfully.
Listener ssl:default on 0.0.0.0:8883 stopped.
Listener tcp:default on 0.0.0.0:1883 stopped.
Listener ws:default on 0.0.0.0:8083 stopped.
Listener wss:default on 0.0.0.0:8084 stopped.
Listener ssl:default on 0.0.0.0:8883 started.
Listener tcp:default on 0.0.0.0:1883 started.
Listener ws:default on 0.0.0.0:8083 started.
Listener wss:default on 0.0.0.0:8084 started.
Listener http:dashboard on :18083 started.
run_erl [2352] Wed Jan 17 11:07:42 2024
Args before exec of shell:
run_erl [2352] Wed Jan 17 11:07:42 2024
argv[0] = sh
run_erl [2352] Wed Jan 17 11:07:42 2024
argv[1] = -c
run_erl [2352] Wed Jan 17 11:07:42 2024
argv[2] = exec "/home/ubuntu/emqx/bin/emqx" "console"
```

</details>

---

**Second Replicant Node** (`emqx@10.1.2.152`)
This was added immediately after adding second core node and has first and second core and first replicant nodes as its static seeds.

Result of `emqx_ctl cluster status`:

```
Cluster status: #{running_nodes =>
                      ['emqx@10.1.1.114','emqx@10.1.1.147','emqx@10.1.2.152',
                       'emqx@10.1.2.222'],
                  stopped_nodes => []}
```

Logs:  

<details>

```
=====
===== LOGGING STARTED Wed Jan 17 11:08:20 UTC 2024
=====
Erlang/OTP 25 [erts-13.2.2] [emqx] [64-bit] [smp:2:2] [ds:2:2:8] [async-threads:4] [jit:ns]

Listener ssl:default on 0.0.0.0:8883 started.
Listener tcp:default on 0.0.0.0:1883 started.
Listener ws:default on 0.0.0.0:8083 started.
Listener wss:default on 0.0.0.0:8084 started.
Listener http:dashboard on :18083 started.
EMQX 5.4.1 is running now!
Restricted Eshell V13.2.2  (abort with ^G)
v5.4.1(emqx@10.1.2.152)1>
===== ALIVE Wed Jan 17 11:23:32 UTC 2024
run_erl [2342] Wed Jan 17 11:08:20 2024
Args before exec of shell:
run_erl [2342] Wed Jan 17 11:08:20 2024
argv[0] = sh
run_erl [2342] Wed Jan 17 11:08:20 2024
argv[1] = -c
run_erl [2342] Wed Jan 17 11:08:20 2024
argv[2] = exec "/home/ubuntu/emqx/bin/emqx" "console"
```

</details>

**Comment by ieQu1** [comment]
Hello,

Thanks for collecting this information. 

1) The following logs indicate that the second core node successfully joined the cluster:

```
v5.4.1(emqx@10.1.2.222)1> 2024-01-17T11:07:54.605274+00:00 [warning] msg: Stopping mria, mfa: mria:stop/1(134), reason: join
Stop listener http:dashboard on :18083 successfully.
Listener ssl:default on 0.0.0.0:8883 stopped.
Listener tcp:default on 0.0.0.0:1883 stopped.
Listener ws:default on 0.0.0.0:8083 stopped.
Listener wss:default on 0.0.0.0:8084 stopped.
Listener ssl:default on 0.0.0.0:8883 started.
Listener tcp:default on 0.0.0.0:1883 started.
Listener ws:default on 0.0.0.0:8083 started.
Listener wss:default on 0.0.0.0:8084 started.
Listener http:dashboard on :18083 started.
```

(Restarting the applications is part of the routine) I don't see any problems here so far.

2) State of the replicant nodes can be further examined with the following commands:

```
emqx eval 'mria_rlog:status()'
```

The output should look like this:

```
# bin/emqx eval 'mria_rlog:status()'
#{backend => rlog,role => replicant,
  shard_stats =>
      #{'$mria_meta_shard' =>
            #{bootstrap_num_keys => 3,bootstrap_time => 2,lag => 0,
              last_imported_trans => 30,message_queue_len => 0,
              replayq_len => 0,state => normal,upstream => 'emqx@n1.local'},

....

  shards_down => [],
  shards_in_sync =>
      ['$mria_meta_shard',emqx_acl_sharded,emqx_authn_shard,
       emqx_cluster_rpc_shard,emqx_cm_shard,emqx_common_shard,
       emqx_dashboard_shard,emqx_ds_builtin_metadata_shard,
       emqx_exclusive_shard,emqx_psk_shard,emqx_retainer_shard,
       emqx_shared_sub_shard,emqx_telemetry_shard,route_shard]}

```

`shards_down => []` part means the replicant is fully in-sync.

Additionally, the following command will give further information about the replicant-core connectivity:

```
# emqx eval 'sys:get_state(mria_lb)'
```

```
# mria_config:core_node_discovery_callback()
```

I expect there to be the records of both core nodes. If not, perhaps there was a network issue, a configuration error, or a genuine bug.

**Comment by seinshah** [comment]
Sorry for the delay.

**Replicant 1 Output**  (which shows the second core node as stopped):

`emqx_ctl cluster status`

<details>

```
Cluster status: #{running_nodes =>
                      ['emqx@10.1.1.196','emqx@10.1.1.220','emqx@10.1.2.139'],
                  stopped_nodes => ['emqx@10.1.2.87']}
```

</details>

`emqx eval 'sys:get_state(mria_lb)'`

<details>

```
{s,['emqx@10.1.1.220'],
   #{'emqx@10.1.1.220' =>
         #{db_nodes => ['emqx@10.1.2.87','emqx@10.1.1.220'],
           protocol_version => 1,running => true,
           shard_badness =>
               [{'$mria_meta_shard',1.94506858418869},
                {emqx_acl_sharded,1.865954367484202},
                {emqx_authn_shard,1.9414674409252854},
                {emqx_cluster_rpc_shard,1.315198400814539},
                {emqx_cm_shard,1.9832858881161037},
                {emqx_common_shard,1.8928669850551976},
                {emqx_dashboard_shard,1.0213865389869206},
                {emqx_exclusive_shard,1.1428801062669889},
                {emqx_psk_shard,1.4489347328171815},
                {emqx_retainer_shard,1.8448336003748467},
                {emqx_shared_sub_shard,1.5973454817154416},
                {emqx_telemetry_shard,1.4823049383312674},
                {route_shard,1.6746019512050414}],
           version => "0.7.1",whoami => core}}}
```

</details>

`emqx eval 'mria_config:core_node_discovery_callback()'`

```
fun ekka_autocluster:core_node_discovery_callback/0
```

----

**Replicant 2 Output** (which works as expected)

`emqx_ctl cluster status`

<details>

```
Cluster status: #{running_nodes =>
                      ['emqx@10.1.1.196','emqx@10.1.1.220','emqx@10.1.2.139',
                       'emqx@10.1.2.87'],
                  stopped_nodes => []}
```

</details>

`emqx eval 'sys:get_state(mria_lb)'`

<details>

```
{s,['emqx@10.1.1.220','emqx@10.1.2.87'],
   #{'emqx@10.1.1.220' =>
         #{db_nodes => ['emqx@10.1.2.87','emqx@10.1.1.220'],
           protocol_version => 1,running => true,
           shard_badness =>
               [{'$mria_meta_shard',1.708136986308328},
                {emqx_acl_sharded,1.4851055760658212},
                {emqx_authn_shard,1.4188610483622486},
                {emqx_cluster_rpc_shard,1.337300887584922},
                {emqx_cm_shard,1.9582848226748935},
                {emqx_common_shard,1.6710703019565953},
                {emqx_dashboard_shard,1.7879678859075816},
                {emqx_exclusive_shard,1.3926508011420278},
                {emqx_psk_shard,1.0281805868664484},
                {emqx_retainer_shard,1.2140955264653406},
                {emqx_shared_sub_shard,1.142949580955571},
                {emqx_telemetry_shard,1.6796434033686019},
                {route_shard,1.4643694238105167}],
           version => "0.7.1",whoami => core},
     'emqx@10.1.2.87' =>
         #{db_nodes => ['emqx@10.1.2.87','emqx@10.1.1.220'],
           protocol_version => 1,running => true,
           shard_badness =>
               [{'$mria_meta_shard',1.322359194091815},
                {emqx_acl_sharded,1.0345998576560342},
                {emqx_authn_shard,1.996853656530615},
                {emqx_cluster_rpc_shard,1.333945223921278},
                {emqx_cm_shard,1.4220965319504464},
                {emqx_common_shard,1.1057458983347952},
                {emqx_dashboard_shard,1.1357916230673748},
                {emqx_exclusive_shard,1.0571346231704641},
                {emqx_psk_shard,1.8889342589104743},
                {emqx_retainer_shard,1.7153988103750164},
                {emqx_shared_sub_shard,1.1566363739380705},
                {emqx_telemetry_shard,1.33662380682925},
                {route_shard,1.599069341120111}],
           version => "0.7.1",whoami => core}}}
```

</details>

`emqx eval 'mria_config:core_node_discovery_callback()'`

```
fun ekka_autocluster:core_node_discovery_callback/0
```

> On dashboard, replicant 1 disappears and appears consistently and core 2 changes to green and red consistently

NOTE: IPs are different from the previous message

**Comment by ieQu1** [comment]
Thanks. It looks like the second replicant wasn't able to reach the second core node for some reason. I will run some local tests to see what can trigger this behavior.

**Comment by seinshah** [comment]
@ieQu1, just to let you know, this issue does not appear when I deploy the core nodes first (core 1 and core 2), and afterward replicant nodes. In this scenario, everything works as expected.

**Comment by ieQu1** [comment]
Thanks for the info.

**Comment by zmstone** [comment]
I suspect the flapping status observed on the dashboard was because the HTTP requests were load-balanced to different nodes in the cluster (maybe even a replicant node).

As observed in the log files, a replicant cannot connect to core2, so they may report core2 as down.
And the ctl command was executed in one particular core node which can see all the nodes as running.
